src/profile/references/ref_macro.rs

```diff
@@ -0,0 +1,173 @@
+/// Declares a `uniffi::Object` exported type holding
+/// an inner type with an `RwLock` so that we can
+/// take the value from an `Arc<Ref>`, thus avoiding
+/// expensive clone operations.
+macro_rules! decl_ref_named {
+    (
+        $struct_name: ty,
+        $inner: ty
+    ) => {
+        paste! {
+            use crate::prelude::*;
+            use std::sync::{Arc, RwLock};
+
+            #[derive(Debug, uniffi::Object)]
+            #[uniffi::export(Debug, Eq, Hash)]
+            pub(crate) struct $struct_name {
+                inner: RwLock<Option<$inner>>,
```

Comment:
This is an EXTREMELY useful trick! We are inside a type which is `uniffi::Object` exported which we use with `Arc`. However, in Rust, an `Arc`:ed type **cannot be mutated**, a method which would otherwise be marked `&mut self` cannot be used. However, we can use **interior mutability**, such as `Cell` / `RefCell`, unfortunately neither of these can be used with UniFFI IIRC. But `RwLock` works with UniFFI. So how do we mutate the inner value without mutating the outer Object? Well we can "take" the `Some($inner)` from the `RwLock` leaving it with `None`, which is OK since it is interior mutability!

Reply:
See `take` method below!